base/math.jl

```diff
@@ -142,16 +144,17 @@ function evalpoly(z::Complex, p::Tuple)
     end
 end
 evalpoly(z::Complex, p::Tuple{<:Any}) = p[1]
-
+evalpoly(z::Complex, ::Tuple{}) = zero(one(z)) # dimensionless zero, i.e. 0 * z^0
 
 evalpoly(z::Complex, p::AbstractVector) = _evalpoly(z, p)
 
 function _evalpoly(z::Complex, p)
     Base.require_one_based_indexing(p)
-    length(p) == 1 && return p[1]
     N = length(p)
-    a = p[end]
-    b = p[end-1]
+    @inbounds p0 = N == 0 ? reduce_empty_iter(+, p) : p[N]
+    N <= 1 && return oftype(one(z) * p0, p0)
```

Comment:
As above, this doesn't help with type stability in general.

Reply:
As I explained above, `one` is the correct function to handle unitful types.